### PR TITLE
Add SQL function overload LOG(base, x) for logarithm of x to base

### LINKED ISSUE #5206: Add SQL function overload `LOG(base, x)` for logarithm of x to base
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

As a user, I would like to determine the logarithm of a value, $x$ to base $b$.

**Describe the solution you'd like**

Add an overload to the `log` function, which takes two arguments. Using PostgreSQL for inspiration per [the documentation](https://www.postgresql.org/docs/current/functions-math.html):

> `log ( b numeric, x numeric ) → numeric`
> Logarithm of $x$ to base $b$
> log(2.0, 64.0) → 6.0000000000000000  

**Describe alternatives you've considered**

DataFusion strives for PostgreSQL compatibility, so it makes sense to use the PostgreSQL solution.

**Additional context**

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #5206 .

# Rationale for this change
Add SQL function overload LOG(base, x) for logarithm of x to base

# What changes are included in this PR?

# Are these changes tested?
Test included

# Are there any user-facing changes?
No

### CODE REVIEW

**Comment by comphead** [comment]
@stuartcarnie @alamb with `log` function we have some discrepancy with PG, its not brought up with this PR, it existed before

```
❯ select log(1, 64), log(0);
+-------------------------+---------------+
| log(Int64(1),Int64(64)) | log(Int64(0)) |
+-------------------------+---------------+
| inf                     | -inf          |
+-------------------------+---------------+
```

whereas PostgresSQL doesn't allow queries like that and fail the query

**Comment by alamb** [comment]
Thank you very much for this contribution @comphead

**Comment by comphead** [comment]
> > whereas PostgresSQL doesn't allow queries like that and fail the query
> 
> For anyone else who is curious, here is what postgres does with this query:
> 
> ```sql
> postgres=# select log(0);
> ERROR:  cannot take logarithm of zero
> postgres=# select log(1, 64);
> ERROR:  division by zero
> ```

@alamb would you like to solve edgecases as part of this PR or another one?

**Comment by alamb** [comment]
> @alamb would you like to solve edgecases as part of this PR or another one?

A different one, please

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks @comphead  -- this looks really great. cc @stuartcarnie

**Comment by stuartcarnie** [comment]
Looks great, thanks @comphead 🙏

File: datafusion/core/tests/sqllogictests/test_files/scalar.slt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+#############
+## Scalar Tests
```
[the remark this answers is not in the document]
Comment (comphead) [reply datafusion/core/tests/sqllogictests/test_files/scalar.slt]: Done

File: datafusion/core/tests/sqllogictests/test_files/scalar.slt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+# log scalar nulls 1
+query IT rowsort
+select log(2, null) a, log(null) b
+----
+NULL NULL
+
+# log scalar nulls 2
+query IT rowsort
+select log(null, null) a, log(null) b
+----
```
[the remark this answers is not in the document]
Comment (comphead) [reply datafusion/core/tests/sqllogictests/test_files/scalar.slt:62]: Done, filed https://github.com/apache/arrow-datafusion/issues/5259

**Comment by alamb** [comment]
Thanks again @comphead

**Pull request merged**